cpp/src/barretenberg/dsl/acir_proofs/acir_proofs.cpp

```diff
@@ -154,4 +189,194 @@ bool verify_proof(
     return verified;
 }
 
+/**
+ * @brief Enables simulation of recursion in the ACVM. In order to be able to simulate execute of ACIR,
+ *        a DSL needs to be able to insert into a circuit's witness the correct output aggregation state of a recursive
+ * verification. The resulting aggergation state is what will be compared against in the recursion constraint.
+ *
+ * @param proof_buf - The proof to be verified
+ * @param proof_length - The length of the proof to enable reading the proof from buffer
+ * @param vk_buf - The verification key of the circuit being verified
+ * @param vk_length - The length of the verification eky to enable reading the proof from buffer
+ * @param public_inputs_buf - The public inputs of the proof being verified. To be included when construct the input
+ * aggregation state
+ * @param input_aggregation_obj_buf - The fields represnting the two G1 points that must be fed into a pairing.
+ *                                    This will be empty if this is the first proof to be recursively.
+ * @param output_aggregation_obj_buf - Two G1 points that the top-level verifier needs to run a pairing upon to complete
+ * verification
+ */
+size_t verify_recursive_proof(uint8_t const* proof_buf,
+                              uint32_t proof_length,
+                              uint8_t const* vk_buf,
+                              uint32_t vk_length,
+                              uint32_t num_public_inputs,
+                              uint8_t const* input_aggregation_obj_buf,
+                              uint8_t** output_aggregation_obj_buf)
+{
+
+    bool inner_aggregation_all_zero = true;
+    std::vector<barretenberg::fr> aggregation_input(16);
+    for (size_t i = 0; i < 16; i++) {
+        aggregation_input[i] = barretenberg::fr::serialize_from_buffer(&input_aggregation_obj_buf[i * 32]);
+        inner_aggregation_all_zero &= (aggregation_input[i].is_zero());
+    }
+
+    acir_format::aggregation_state_ct previous_aggregation;
+    if (!inner_aggregation_all_zero) {
+        std::array<acir_format::bn254::fq_ct, 4> aggregation_elements;
```

Comment:
what does 4 represent here?

Reply:
It represents the total number of field elements needed to represent an aggregation state. I'm going to see how I can switch this to use constexpr's and avoid the magic numbers

Reply:
I introduced a new `NUM_AGGREGATION_ELEMENTS` in `recursion_constraint.hpp`. Let me know what you think and if I missed anywhere that it can be used